### PR TITLE
fix(STONEINTG-565): change the condition for HaveBindingsFailed

### PR DESCRIPTION
**Pull request by dirgim**
* The HaveBindingsFailed function was using the wrong condition ComponentDeploymentConditions, changed it to use BindingConditions
* Add additional checks to ensure that EnsureIntegrationTestPipelineForScenarioExists executes correctly
* Add IntegrationSnapshotEnvironmentBindingPredicate to the binding controller to ensure that individual predicates don't filter SEBs that are not for Integration testing
* Cover more of gitops/binding.go with dedicated file
* Add test coverage for updated IsBindingDeployed and HaveBindingsFailed functions

## Maintainers will complete the following section

- [x] Commit messages are descriptive enough ([hints](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/))
- [x] Code coverage from testing does not decrease and new code is covered
- [x] [Controllers diagrams](https://github.com/redhat-appstudio/integration-service/tree/main/docs) are updated when PR changes controllers code  (if applicable)

### CODE REVIEW

File: controllers/binding/snapshotenvironmentbinding_adapter.go [hunk 1]
```diff
@@ -108,14 +114,14 @@ func (a *Adapter) EnsureIntegrationTestPipelineForScenarioExists() (controller.O
 // EnsureEphemeralEnvironmentsCleanedUp will ensure that ephemeral environment(s) associated with the
 // SnapshotEnvironmentBinding are cleaned up.
 func (a *Adapter) EnsureEphemeralEnvironmentsCleanedUp() (controller.OperationResult, error) {
-
 	if !gitops.HaveBindingsFailed(a.snapshotEnvironmentBinding) {
 		return controller.ContinueProcessing()
 	}
+	a.logger.Info("The SnapshotEnvironmentBinding encountered an issue deploying snapshot on ephemeral environments", "snapshotEnvironmentBinding.Name", a.snapshotEnvironmentBinding.Name)
```
Comment (MartinBasti) [inline controllers/binding/snapshotenvironmentbinding_adapter.go:120]: Can we log, what was the issue?

File: gitops/binding_test.go [hunk 2]
```diff
@@ -3,14 +3,13 @@ package gitops_test
 import (
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/types"
```
Comment (MartinBasti) [inline gitops/binding_test.go:6]: Why this change? It looks like fight of IDEs, where one moves it to top and other back to bottom of this section.

**Review by Josh-Everett — APPROVED** [review]

**Review by dheerajodha — APPROVED** [review]
LGTM!

**Comment by dheerajodha** [comment]
@dirgim weird how it always fails on `s: "tests haven't succeeded yet for snapshot integration-e2e-tenant/snapshot-sample-ygfk",`

**Review by hongweiliu17 — APPROVED** [review]
LGTM

File: controllers/binding/snapshotenvironmentbinding_adapter.go [hunk 3]
```diff
@@ -108,14 +114,16 @@ func (a *Adapter) EnsureIntegrationTestPipelineForScenarioExists() (controller.O
...
 func (a *Adapter) EnsureEphemeralEnvironmentsCleanedUp() (controller.OperationResult, error) {
-
 	if !gitops.HaveBindingsFailed(a.snapshotEnvironmentBinding) {
 		return controller.ContinueProcessing()
 	}
 
 	// mark snapshot as failed
 	snapshotErrorMessage := "Encountered issue deploying snapshot on ephemeral environments: " +
-		meta.FindStatusCondition(a.snapshotEnvironmentBinding.Status.ComponentDeploymentConditions, "ErrorOccurred").Message
+		meta.FindStatusCondition(a.snapshotEnvironmentBinding.Status.BindingConditions, "ErrorOccurred").Message
```
[the remark this answers is not in the document]
Comment (dirgim) [reply controllers/binding/snapshotenvironmentbinding_adapter.go:123]: In my testing it happened that sometimes `ComponentDeploymentConditions` weren't even present on the SEB, causing an error on this line.
I would expect the `BindingConditions` to have relevant information in any case.

File: docs/binding-controller.md [hunk 4]
```diff
@@ -6,6 +6,7 @@ flowchart TD
     classDef Amber fill:#FFDEAD;
     classDef Green fill:#BDFFA4;
 
+predicate_integration_seb((PREDICATE: <br>SnapshotEnvironmentBinding<br>is associated with<br>IntegrationTestScenario))
```
[the remark this answers is not in the document]
Comment (dirgim) [reply docs/binding-controller.md:9]: Ack, missed that - added now.

File: gitops/predicates.go [hunk 5]
```diff
@@ -19,13 +19,13 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 )
 
-// DeploymentFinishedForIntegrationBindingPredicate returns a predicate which filters out update events to a
+// DeploymentSucceededForIntegrationBindingPredicate returns a predicate which filters out update events to a
 // SnapshotEnvironmentBinding that has the IntegrationTestScenario specified in the label and
 // where the component deployment status goes from unknown to true/false.
```
[the remark this answers is not in the document]
Comment (dirgim) [reply gitops/predicates.go]: Good point, updated the description.

File: gitops/predicates.go [hunk 6]
```diff
@@ -39,3 +39,13 @@ func DeploymentFailedForIntegrationBindingPredicate() predicate.Predicate {
 		},
 	}
 }
+
+// IntegrationSnapshotEnvironmentBindingPredicate returns a predicate which filters out update events to a
+// SnapshotEnvironmentBinding is associated with an IntegrationTestScenario.
```
[the remark this answers is not in the document]
Comment (dirgim) [reply gitops/predicates.go]: Ack, updated - I caught it when I realized that we were picking up events for the default `development` environment.

File: controllers/binding/snapshotenvironmentbinding_adapter.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dheerajodha) [reply controllers/binding/snapshotenvironmentbinding_adapter.go:123]: Thanks for the info!

File: gitops/predicates.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dheerajodha) [reply gitops/predicates.go]: That's very nice, thanks!

File: gitops/predicates_test.go [hunk 7]
```diff
@@ -168,4 +168,57 @@ var _ = Describe("Predicates", Ordered, func() {
...
+		It("returns false when the SEB with SnapshotTestScenarioLabel is deleted", func() {
+			bindingTrueStatus.ObjectMeta.Labels = map[string]string{}
+
+			contextEvent := event.DeleteEvent{
+				Object: bindingMissingStatus,
+			}
+			Expect(instance.Delete(contextEvent)).To(BeFalse())
+		})
+		It("returns false when the SEB with SnapshotTestScenarioLabel encounters a generic event", func() {
+			bindingTrueStatus.ObjectMeta.Labels = map[string]string{}
```
Comment (dheerajodha) [inline gitops/predicates_test.go:216]: missing `SnapshotTestScenarioLabel`

Comment (dirgim) [reply gitops/predicates_test.go]: Good catch, updated.

**Review by jsztuka — APPROVED** [review]
lgtm
